java/core/src/main/java/com/suse/manager/reactor/messaging/JobReturnEventMessageAction.java

```diff
@@ -276,31 +185,128 @@ else if (packageRefreshNeeded.contains(PackageChangeOutcome.NEEDS_REFRESHING)) {
+        try {
+            actionChainResult = Json.GSON.fromJson(jsonResult,
+                    new TypeToken<Map<String, StateApplyResult<Ret<JsonElement>>>>() {
+                    }.getType());
+        }
+        catch (JsonSyntaxException e) {
+            LOG.error("Error mapping action chain result: {}", jsonResult, e);
+            throw e;
+        }
+
+        saltServerActionService.handleActionChainResult(jobReturnEvent.getMinionId(), jobReturnEvent.getJobId(),
+                actionChainResult, stateResult -> false);
+
+        if (isFunctionTestMode) {
+            return;
+        }
+
+        Set<PackageChangeOutcome> packageRefreshNeeded = actionChainResult.entrySet().stream()
+                .map(entry -> SaltActionChainGeneratorService.parseActionChainStateId(entry.getKey())
+                        .map(stateId -> handlePackageChanges(jobReturnEvent, entry.getValue().getName(),
+                                Optional.ofNullable(entry.getValue().getChanges().getRet())))
+                        .orElse(PackageChangeOutcome.DONE)
+                ).collect(Collectors.toSet());
+
+        if (packageRefreshNeeded.contains(PackageChangeOutcome.NEEDS_DELAYED_REFRESHING)) {
+            schedulePackageRefresh(scheduler.get(), jobReturnEvent.getMinionId(),
+                    Date.from(Instant.now().plusSeconds(60)));
+        }
+        else if (packageRefreshNeeded.contains(PackageChangeOutcome.NEEDS_REFRESHING)) {
+            schedulePackageRefresh(scheduler.get(), jobReturnEvent.getMinionId());
+        }
+    }
+
+    private void handleNotActionChainNorTestMode(Optional<JsonElement> jobResult, JobReturnEvent jobReturnEvent,
+                                                 String function, Optional<Long> actionId) {
+        Date earliest = new Date();
+        PackageChangeOutcome changeOutcome = handlePackageChanges(jobReturnEvent,
+                Optional.ofNullable(function).map(Xor::right), jobResult);
+        if (changeOutcome != PackageChangeOutcome.DONE) {
+            if (changeOutcome == PackageChangeOutcome.NEEDS_DELAYED_REFRESHING) {
+                earliest = Date.from(Instant.now().plusSeconds(60));
+            }
+            Optional<User> scheduler = Optional.empty();
+            if (actionId.isPresent()) {
+                Optional<Action> action = Optional.ofNullable(ActionFactory.lookupById(actionId.get()));
+                if (action.isPresent()) {
+                    scheduler = Optional.ofNullable(action.get().getSchedulerUser());
+                    if (action.get().getActionType().equals(ActionFactory.TYPE_DIST_UPGRADE)) {
+                        Calendar calendar = Calendar.getInstance();
+                        calendar.add(Calendar.SECOND, 30);
+                        earliest = calendar.getTime();
+                    }
                 }
             }
+            schedulePackageRefresh(scheduler, jobReturnEvent.getMinionId(), earliest);
```

Comment:
Not your change of course, but this part is kind of a mess due to the usage of `Optional`s and the fact that they cannot be combined due to `earliest` and `scheduler` not being final. Moreover, if the outcome  is `NEEDS_DELAYED_REFRESHING` but type is `TYPE_DIST_UPGRADE`, the wait is 30 seconds instead of 60, not sure if this is intended or a bug.